### PR TITLE
support slicetransform

### PR DESCRIPTION
**Pull request by siddontang**
@BusyJay @zhangjinpeng1987

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
When we use prefix extractor, we need set `whole_key_filtering` to false. Please add `rocksdb_block_based_options_set_whole_key_filtering`.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/compaction_filter.rs [hunk 1]
```diff
@@ -1,6 +1,7 @@
-use libc::{c_void, c_char, c_int, size_t};
+
```
Comment (BusyJay) [inline src/compaction_filter.rs]: Why keep these two empty lines?

File: src/rocksdb.rs [hunk 2]
```diff
@@ -115,7 +115,9 @@ impl<'a> DBIterator<'a> {
                 SeekKey::Start => crocksdb_ffi::crocksdb_iter_seek_to_first(self.inner),
                 SeekKey::End => crocksdb_ffi::crocksdb_iter_seek_to_last(self.inner),
                 SeekKey::Key(key) => {
-                    crocksdb_ffi::crocksdb_iter_seek_for_prev(self.inner, key.as_ptr(), key.len() as size_t)
+                    crocksdb_ffi::crocksdb_iter_seek_for_prev(self.inner,
```
Comment (BusyJay) [inline src/rocksdb.rs]: I propose to raise a pr to format the code first.

File: src/slice_transform.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                        key: *const u8,
+                        key_len: size_t,
+                        dest_len: *mut size_t)
+                        -> *const u8 {
+    unsafe {
+        let transform = &mut *(transform as *mut SliceTransformProxy);
+        let key = slice::from_raw_parts(key, key_len);
+        let prefix = transform.transform.transform(key);
+        *dest_len = prefix.len() as size_t;
+        return prefix.as_ptr() as *const u8;
```
Comment (BusyJay) [inline src/slice_transform.rs]: `return` is unnecessary.

File: test/test_slice_transform.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+use rocksdb::{Writable, DB, SliceTransform, Options, SeekKey, BlockBasedOptions};
+use tempdir::TempDir;
+
+struct FixedPostfixTransform {
+    pub postfix_len: usize,
+}
+
+impl SliceTransform for FixedPostfixTransform {
+    fn transform<'a>(&mut self, key: &'a [u8]) -> &'a [u8] {
+        let mid = key.len() - self.postfix_len;
```
Comment (BusyJay) [inline test/test_slice_transform.rs]: Why not just `&key[..self.postfix_len]`?

File: src/slice_transform.rs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    // specified in the Transform method. This method is invoked for every
+    // key that is inserted into the db. If this method returns true,
+    // then Transform is called to translate the key to its prefix and
+    // that returned prefix is inserted into the bloom filter. If this
+    // method returns false, then the call to Transform is skipped and
+    // no prefix is inserted into the bloom filters.
+    fn in_domain(&mut self, key: &[u8]) -> bool;
+
+    // This is currently not used and remains here for backward compatibility.
+    fn in_range(&mut self, key: &[u8]) -> bool;
```
Comment (BusyJay) [inline src/slice_transform.rs]: We can provide a default implementation.

File: test/test_slice_transform.rs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    ];
+
+    for &(ref k, ref v) in &samples {
+        db.put(k, v).unwrap();
+        assert_eq!(v.as_slice(), &*db.get(k).unwrap().unwrap());
+    }
+
+    let mut it = db.iter();
+
+    let invalid_seeks = vec![
```
Comment (BusyJay) [inline test/test_slice_transform.rs]: Why are these keys invalid?

Comment (siddontang) [inline test/test_slice_transform.rs]: We will remove two postfix bytes for the key and then insert it to the bloom filter, and these keys can't be found in the bloom filter.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/slice_transform.rs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        *dest_len = prefix.len() as size_t;
+        prefix.as_ptr() as *const u8
+    }
+}
+
+extern "C" fn in_domain(transform: *mut c_void, key: *const u8, key_len: size_t) -> u8 {
+    unsafe {
+        let transform = &mut *(transform as *mut SliceTransformProxy);
+        let key = slice::from_raw_parts(key, key_len);
+        return transform.transform.in_domain(key) as u8;
```
Comment (BusyJay) [inline src/slice_transform.rs]: `return` is unnecessary.

File: src/slice_transform.rs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    }
+}
+
+extern "C" fn in_range(transform: *mut c_void, key: *const u8, key_len: size_t) -> u8 {
+    unsafe {
+        let transform = &mut *(transform as *mut SliceTransformProxy);
+        let key = slice::from_raw_parts(key, key_len);
+        return transform.transform.in_range(key) as u8;
```
Comment (BusyJay) [inline src/slice_transform.rs]: `return` is unnecessary.

File: test/test_slice_transform.rs (hunk 4 above)
Comment (siddontang) [inline test/test_slice_transform.rs]: it is postfix, not prefix.

Comment (siddontang) [inline test/test_slice_transform.rs]: Assume the key length is 10, and postfix length is 2, we will remove the last 2 bytes and extract the first 8 bytes as the key.

**Comment by siddontang** [comment]
PTAL @BusyJay

**Pull request merged**